bfvmm/src/hve/arch/intel_x64/vmi/vmi_op.cpp

```diff
@@ -0,0 +1,185 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <hve/arch/intel_x64/vcpu.h>
+#include <hve/arch/intel_x64/emulation/cpuid.h>
+#include <hve/arch/intel_x64/vmi/vmi_op.h>
+
+#include <bfhypercall.h>
+#include <bfjson.h>
+
+#include <algorithm>
+
+namespace boxy::intel_x64
+{
+
+bool vmi_op_vmcall_dispatcher (vcpu_t *vcpu);
+
+vmi_op_handler::vmi_op_handler(
+    gsl::not_null<vcpu *> vcpu
+) :
+    m_vcpu{vcpu}
+{
+
+    // TODO:
+    //
+    // For now, dom0 is always the target of the introspecting VM
+    // and domU is always the introspecting VM.
+    // This will change once the vmi policy_op is implemented.
+
+    if(vcpu->is_domU()) {
+        vcpu->add_vmcall_handler({&vmi_op_handler::dispatch, this});
+        vcpu->add_cpuid_emulator(
+            bf_vmi_cpuid_leaf, {&vmi_op_handler::cpuid_vmi, this});
+    }
+}
+
+// Note:
+//
+// LibVMI currently makes hypercalls without changing the native ABI, except for setting RAX to the
+// hypercall opcode. Therefore, unlike other handlers, we have to follow SystemV's AMD64 ABI for
+// the calling convention to handle them.
+// As a reminder, the order is: RDI, RSI, RDX, RCX, R8, R9.
+
+void
+vmi_op_handler::vmi_op__memmap_ept(vcpu *vcpu)
+{
+    uint64_t addr = vcpu->rdi();
+    uint64_t gpa2 = vcpu->rsi();
+
+    auto [gpa1, unused1] = vcpu->gva_to_gpa(addr);
+    auto &&guest_map = get_domain(vcpu->domid())->ept();
+
+    if(guest_map.is_2m(gpa1))
+    {
+        bfdebug_info(0, "vmi_op: guest_map is 2m");
+        auto gpa1_2m = bfn::upper(gpa1, ::intel_x64::ept::pd::from);
+        bfvmm::intel_x64::ept::identity_map_convert_2m_to_4k(guest_map, gpa1_2m);
+    }
+
+    auto gpa1_4k = bfn::upper(gpa1, ::intel_x64::ept::pt::from);
+    auto gpa2_4k = bfn::upper(gpa2, ::intel_x64::ept::pt::from);
+
+    vcpu->set_rsi(gpa2_4k);
+
+    auto [pte, unused2] = guest_map.entry(gpa1_4k);
+    ::intel_x64::ept::pt::entry::phys_addr::set(pte, gpa2_4k);
+
+    // flush EPT tlb, guest TLB doesn't need to be flushed
+    // as that translation hasn't changed
+    ::intel_x64::vmx::invept_global();
+}
+
+void
+vmi_op_handler::vmi_op__translate_v2p(vcpu *vcpu)
+{
+    auto addr = vcpu->rdi();
+    auto [gpa, unused] = vcpu->gva_to_gpa(addr);
+    vcpu->set_rdi(gpa);
+}
+
+void
+vmi_op_handler::vmi_op__get_register_data(vcpu *vcpu)
+{
+    uintptr_t addr = vcpu->rdi();
+    uint64_t size = vcpu->rsi();
+    json j;
+
+    // TODO: Find the target vcpu once policy_op is implemented
+    auto target_vcpu = vcpu->parent_vcpu();
+    target_vcpu->load();
+
+    j["RAX"] = target_vcpu->rax();
```

Comment:
Is there a way to remove the need for JSON here? I understand why we did this originally, but there is a huge issue with the use of JSON. Specifically, it performs allocations in the hypervisor which results in fragmentation. Allocations should never happen from a exit handler (including VMCalls) outside of doing things like creating new vCPUs and domains. A versioned structure makes a lot more sense here. It would be faster, and it would remove this long-term issue with the ABI. It also removes the need for an additional library

Reply:
Would using a library like https://github.com/beached/daw_json_link help? It advertises itself as a *Static JSON parsing in C++* library and seems to use `constexpr` everywhere. I can do some testing to see how it handles (de)serialization and large values.

Reply:
I can also delay the need to use an external JSON library allowing us to discuss this further by implementing a hypercall for each register which is what [domain_op](https://github.com/Bareflank/boxy/blob/872d0afb55a0f48f2cf49f783e7b39b2728f7744/bfvmm/src/hve/arch/intel_x64/vmcall/domain_op.cpp) is already doing.

It might be worth taking some time before deciding to add support for an external library as Bareflank is currently making efforts to be AUTOSAR and C++ Core Guideline compliant with the BSL. My understand is that future version of Bareflank will need to be AUTOSAR compliant in the VMM and adding support for an external library now might delay the support of upcoming Bareflank VMMs with LibVMI.

Reply:
The conversation is continuing on [slack #boxy](https://bareflank.slack.com/archives/CUB38DB4J/p1593009311057400)